backend/pkg/notification/queuing.go

```
@@ -424,6 +424,7 @@ func RenderEmailsForUserEvents(epoch uint64, notificationsByUserID types.Notific
 
 func QueueEmailNotifications(epoch uint64, notificationsByUserID types.NotificationsPerUserId, tx *sqlx.Tx) error {
 	// for emails multiple notifications will be rendered to one email per user for each run
+	// TODO filter out ratelimited users, don't need to queue or even render emails
 	emails, err := RenderEmailsForUserEvents(epoch, notificationsByUserID)
 	if err != nil {
```

This would be the clean thing to do. Instead of storing rendered emails to db first and let the sender throw them away again later, doing it here with batch querying of user premium perks + email counts from redis would be a nice early out.